nixos/modules/misc/nixpkgs.nix

```diff
@@ -324,74 +324,135 @@ in
-            then config.nixpkgs.crossSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.crossSystem.config))
-            else config.nixpkgs.localSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.localSystem.config));
-          nixosOption =
-            if config.nixpkgs.crossSystem != null
-            then "nixpkgs.crossSystem"
-            else "nixpkgs.localSystem";
-          pkgsSystem = finalPkgs.stdenv.targetPlatform.system;
-        in {
-          assertion = constructedByMe -> !hasPlatform -> nixosExpectedSystem == pkgsSystem;
-          message = "The NixOS nixpkgs.pkgs option was set to a Nixpkgs invocation that compiles to target system ${pkgsSystem} but NixOS was configured for system ${nixosExpectedSystem} via NixOS option ${nixosOption}. The NixOS system settings must match the Nixpkgs target system.";
-        }
-      )
-      {
-        assertion = constructedByMe -> hasPlatform -> legacyOptionsDefined == [];
-        message = ''
-          Your system configures nixpkgs with the platform parameter${optionalString hasBuildPlatform "s"}:
-          ${hostPlatformLine
-          }${buildPlatformLine
-          }
-          However, it also defines the legacy options:
-          ${concatMapStrings showOptionWithDefLocs legacyOptionsDefined}
-          For a future proof system configuration, we recommend to remove
-          the legacy definitions.
+        description = mdDoc ''
+          The derivation to the nixpkgs sources, as a string with context. This
+          is set up by the nixpkgs flake.
+        '';
+      };
+
+      setNixPath = mkOption {
+        type = types.bool;
+
+        default = cfg.flake.source != null;
+        defaultText = "config.nixpkgs.flake.source != null";
+
+        description = mdDoc ''
+          Whether to create `/etc/nix/inputs/nixpkgs` as a symlink to the version of nixpkgs that
+          the NixOS system was built with and set `NIX_PATH` to include `/etc/nix/inputs/nixpkgs`.
+
+          This makes {command}`nix-build '<nixpkgs>' -A hello` work out of the box on flake systems.
         '';
-      }
-      {
-        assertion = opt.pkgs.isDefined -> cfg.config == {};
-        message = ''
-          Your system configures nixpkgs with an externally created instance.
-          `nixpkgs.config` options should be passed when creating the instance instead.
-
-          Current value:
-          ${lib.generators.toPretty { multiline = true; } opt.config}
+      };
+
+      setFlakeRegistry = mkOption {
+        type = types.bool;
+
+        default = cfg.flake.source != null;
+        defaultText = "config.nixpkgs.flake.source != null";
+
+        description = mdDoc ''
+          Whether to pin nixpkgs in the system-wide flake registry (`/etc/nix/registry.json`) to the
+          version of nixpkgs that the NixOS system has been built with.
```

Comment:
so the understanding I get from this is that it replicates the protocol, URL, etc but pins based on rev. correct? I still feel like there could be a bit more wording to say so explicitly, but it's good as-is.

Reply:
Nope. It is a string-with-context that introduces a simple dependency: 

```
nix-repl> nixosConfigurations.micro.options.nix.registry.value.nixpkgs.to
{ path = "/nix/store/c8cz4q0y6jw9qvjwk45nf6pvwysj7yn6-source"; type = "path"; }

nix-repl> :p builtins.getContext nixosConfigurations.micro.options.nix.registry.value.nixpkgs.to.path
{ "/nix/store/c8cz4q0y6jw9qvjwk45nf6pvwysj7yn6-source" = { path = true; }; }
```

Reply:
ah! that's nice and simple. the clearest way to say that might be "by referencing it in the store"?